### PR TITLE
Added edit user, change password pages with full functionality and he…

### PR DESCRIPTION
**Pull request by SergeyGhazaryan**

### CODE REVIEW

File: api/src/GTAVersions.Domain/GTAVersions.Domain.csproj [hunk 1]
```diff
@@ -8,6 +8,7 @@
     <PackageReference Include="EnterpriseLibrary.Common" Version="6.0.1304" />
     <PackageReference Include="IdentityServer4.Storage" Version="4.1.2" />
     <PackageReference Include="Mapster" Version="7.2.0" />
+    <PackageReference Include="Microsoft.AspNet.Identity.EntityFramework" Version="2.2.3" />
```
Comment (arayik-99) [inline api/src/GTAVersions.Domain/GTAVersions.Domain.csproj]: Why do we need this?

File: api/src/GTAVersions.Domain/Services/AuthService.cs [hunk 2]
```diff
@@ -34,7 +34,7 @@ public async Task<AccessToken> Login(SignInUserDTO request)
                 throw new HttpResponseException(HttpStatusCode.Unauthorized);
             }
 
-            var token = await _userService.UpdateAndReturnUserTokenAsync(userDTO);
+            var token = await _userService.UpdateAndReturnUserToken(userDTO);
```
Comment (arayik-99) [inline api/src/GTAVersions.Domain/Services/AuthService.cs]: Why was the name changed?

File: api/src/GTAVersions.Domain/Services/UserService.cs [hunk 3]
```diff
@@ -44,11 +44,36 @@ public async Task<AccessToken> UpdateAndReturnUserTokenAsync(UserDTO userDTO)
...
-            var userid = await _userRepository.Create(username, passwordHash);
+            var userid = await _userRepository.CreateUser(firstName, lastName, username, passwordHash);
 
             return userid;
         }
+
+        public async Task<UserDTO> EditUser(EditUserDTO editUserDTO, string currentUserId)
+        {
+            var id = int.Parse(currentUserId);
+
```
Comment (arayik-99) [inline api/src/GTAVersions.Domain/Services/UserService.cs]: Redudant line.

File: api/src/GTAVersions.Domain/Services/UserService.cs [hunk 4]
```diff
@@ -44,11 +44,36 @@ public async Task<AccessToken> UpdateAndReturnUserTokenAsync(UserDTO userDTO)
...
+        {
+            var changedPassword = await _userRepository.ChangePassword(id, passwordHash);
+
+            return changedPassword.Adapt<UserDTO>();
+        }
+
+        public async Task<UserDTO> GetCurrentUser(string currentUserid)
+        {
+            var id = int.Parse(currentUserid);
+
```
Comment (arayik-99) [inline api/src/GTAVersions.Domain/Services/UserService.cs]: Redudant line.

File: api/src/GTAVersions.Domain/Services/AuthService.cs [hunk 5]
```diff
@@ -43,18 +43,38 @@ public async Task<AccessToken> Signup(SignUpUserDTO request)
...
+
+            var token = await _userService.UpdateAndReturnUserToken(createdUser);
 
             return token;
         }
+
+        public async Task<UserDTO> ChangePassword(ChangePasswordDTO request, string currentUserId)
+        {
+            var id = int.Parse(currentUserId);
+
```
Comment (arayik-99) [inline api/src/GTAVersions.Domain/Services/AuthService.cs]: Redudant line.

File: api/src/GTAVersions.Domain/Services/AuthService.cs [hunk 6]
```diff
@@ -43,18 +43,38 @@ public async Task<AccessToken> Signup(SignUpUserDTO request)
...
+            var currentUser = await _userService.GetUserById(id);
+
```
Comment (arayik-99) [inline api/src/GTAVersions.Domain/Services/AuthService.cs:55]: Redudant line.

File: api/src/GTAVersions.Domain/Services/AuthService.cs [hunk 7]
```diff
@@ -43,18 +43,38 @@ public async Task<AccessToken> Signup(SignUpUserDTO request)
...
+            var verified = _passwordHasher.Check(currentUser.PasswordHash, request.OldPassword);
+
+            if (!verified)
+            {
+                throw new HttpResponseException(HttpStatusCode.Unauthorized);
+            }
+
+            var passwordHash = _passwordHasher.PasswordHash(request.NewPassword);
+
```
Comment (arayik-99) [inline api/src/GTAVersions.Domain/Services/AuthService.cs]: Redudant line.

File: api/src/GTAVersions/Controllers/AuthController.cs [hunk 8]
```diff
@@ -31,5 +34,16 @@ public async Task<IActionResult> Signup([FromBody] SignUpUserDTO model)
 
             return Ok(token);
         }
+
+        [Authorize]
+        [HttpPut("settings")]
```
Comment (arayik-99) [inline api/src/GTAVersions/Controllers/AuthController.cs]: Rename this changePassword and change method to PATCH (Read difference between PUT and PATCH)

File: api/src/GTAVersions/Controllers/UserController.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        public UserController(IUserService userService)
+        {
+            _userService = userService;
+        }
+
+        [HttpPut("me/settings")]
+        public async Task<IActionResult> EditUser([FromBody] EditUserDTO editUserDTO)
+        {
+            var currentUserId = HttpContext.User.FindFirstValue(JwtRegisteredClaimNames.Sub);
+
```
Comment (arayik-99) [inline api/src/GTAVersions/Controllers/UserController.cs:28]: Redudant line.

File: api/src/GTAVersions/Controllers/UserController.cs [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+            var editedUser = await _userService.EditUser(editUserDTO, currentUserId);
+
+            return Ok(editedUser);
+        }
+
+        [HttpGet("me")]
+        public async Task<IActionResult> GetCurrentUser()
+        {
+            var currentUserId = HttpContext.User.FindFirstValue(JwtRegisteredClaimNames.Sub);
+
```
Comment (arayik-99) [inline api/src/GTAVersions/Controllers/UserController.cs:37]: Redudant line.

**Review by arayik-99 — CHANGES_REQUESTED** [review]
Header should be used in one place, at the top of the program, not imported in every page.

File: api/src/GTAVersions.Domain/Services/AuthService.cs (hunk 2 above)
Comment (SergeyGhazaryan) [reply api/src/GTAVersions.Domain/Services/AuthService.cs]: Multiple places with async, multiple places without async. I decided to remove them.

File: api/src/GTAVersions.Data/Interfaces/IUserRepository.cs [hunk 11]
```diff
@@ -7,6 +7,9 @@ public interface IUserRepository
     {
         Task<User> GetUserByUsername(string username);
         Task<User> GetUserById(int id);
-        Task<int> Create(string username, string passwordHash);
+        Task<int> CreateUser(string firstName, string lastName, string username, string passwordHash);
+        Task<User> EditUser(string firstName, string lastName, string username, int id);
```
Comment (Shitikyan) [inline api/src/GTAVersions.Data/Interfaces/IUserRepository.cs]: make id as first parameter

File: api/src/GTAVersions.Data/Repositories/UserRepository.cs [hunk 12]
```diff
@@ -29,12 +29,36 @@ public async Task<User> GetUserById(int id)
             return result;
         }
 
-        public async Task<int> Create(string username, string passwordHash)
+        public async Task<int> CreateUser(string firstName, string lastName, string username, string passwordHash)
         {
-            var query = "INSERT INTO Users (Username, PasswordHash) OUTPUT INSERTED.Id VALUES (@Username, @PasswordHash)";
-            var param = new { Username = username, PasswordHash = passwordHash };
+            var query = "INSERT INTO Users (FirstName, LastName, Username, PasswordHash) OUTPUT INSERTED.Id VALUES (@FirstName, @LastName, @Username, @PasswordHash)";
```
Comment (Shitikyan) [inline api/src/GTAVersions.Data/Repositories/UserRepository.cs:34]: no need @PasswordHash for output

File: api/src/GTAVersions.Domain/Interfaces/IUserService.cs [hunk 13]
```diff
@@ -6,9 +6,12 @@ namespace GTAVersions.Domain.Interfaces
...
     {
-        Task<UserDTO> GetUserByUsernameAsync(string username);
-        Task<UserDTO> GetUserByIdAsync(int id);
-        Task<AccessToken> UpdateAndReturnUserTokenAsync(UserDTO userDTO);
-        Task<int> CreateUserAsync(string username, string passwordHash);
+        Task<UserDTO> GetUserByUsername(string username);
+        Task<UserDTO> GetUserById(int id);
+        Task<AccessToken> UpdateAndReturnUserToken(UserDTO userDTO);
+        Task<int> CreateUser(string name, string surname, string username, string passwordHash);
+        Task<UserDTO> EditUser(EditUserDTO editUserDTO, string currentUserId);
```
Comment (Shitikyan) [inline api/src/GTAVersions.Domain/Interfaces/IUserService.cs]: why user id is string?

Comment (Shitikyan) [reply api/src/GTAVersions.Domain/Interfaces/IUserService.cs]: make it as first param

File: gta-gallery/src/components/changePassword/index.jsx [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    ) {
+      await changePassword(newPasswordValue, oldPasswordValue);
+      history.push('/');
+    }
+    return;
+  };
+
+  return (
+    <>
+      <Header />
```
Comment (Shitikyan) [inline gta-gallery/src/components/changePassword/index.jsx]: not correct behavior of using Header

File: gta-gallery/src/components/editUser/index.jsx [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    if (usernameValue && firstNameValue && lastNameValue) {
+      await editUser(firstNameValue, lastNameValue, usernameValue);
+      history.push('/');
+    }
+    return;
+  };
+
+  return (
+    <>
+      <Header />
```
Comment (Shitikyan) [inline gta-gallery/src/components/editUser/index.jsx]: not correct behavior of using Header

File: gta-gallery/src/components/gtaVersionDetails/index.jsx [hunk 16]
```diff
@@ -26,26 +27,28 @@ export const GTAVersionDetails = () => {
...
-            variant='contained'
-            text='UPDATE VERSION'
-          />
-          {updateModalVisible && (
-            <UpdateVersionModal
-              id={id}
-              isOpen={updateModalVisible}
-              onCancel={toggleUpdateDialog}
+    <>
+      <Header />
```
Comment (Shitikyan) [inline gta-gallery/src/components/gtaVersionDetails/index.jsx]: not correct behavior of using Header

File: gta-gallery/src/layout/index.jsx [hunk 17]
```diff
@@ -27,20 +25,14 @@ const Layout = () => {
...
 
-  const handleLogout = () => {
-    localStorage.setItem('token', '');
-    history.push('/login');
-  };
-
   return (
     <>
-      <div className='header'>GTA Versions</div>
+      <Header />
```
Comment (Shitikyan) [inline gta-gallery/src/layout/index.jsx]: not correct behavior of using Header

File: api/src/GTAVersions/Controllers/UserController.cs [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+        public async Task<IActionResult> EditUser([FromBody] EditUserDTO editUserDTO)
+        {
+            var currentUserId = int.Parse(HttpContext.User.FindFirstValue(JwtRegisteredClaimNames.Sub));
+            var editedUser = await _userService.EditUser(currentUserId, editUserDTO);
+
+            return Ok(editedUser);
+        }
+
+        [HttpGet("me")]
+        public async Task<IActionResult> GetCurrentUser()
```
Comment (arayik-99) [inline api/src/GTAVersions/Controllers/UserController.cs]: Move this method upper. The controllers are usually in this order - GET, POST, PUT/PATCH, DELETE

File: gta-gallery/src/components/header/index.jsx [hunk 19]
```diff
@@ -0,0 +1,88 @@
+import React, { useState, useEffect, useMemo } from 'react';
+import { useHistory } from 'react-router-dom';
+import { useSelector, useDispatch } from 'react-redux';
+import { EditOutlined, LogoutOutlined, UserOutlined } from '@ant-design/icons';
+import { Menu, Dropdown } from 'antd';
```
Comment (arayik-99) [inline gta-gallery/src/components/header/index.jsx:3]: Move this up

File: gta-gallery/src/components/header/index.jsx [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+import { getCurrentUser } from '../../services/userService';
+import { isEmpty } from 'lodash';
```
Comment (arayik-99) [inline gta-gallery/src/components/header/index.jsx:6]: move this after react-redux

File: gta-gallery/src/components/login/index.jsx [hunk 21]
```diff
@@ -1,15 +1,18 @@
 import React, { useState } from 'react';
 import { useHistory } from 'react-router-dom';
+import { useDispatch } from 'react-redux';
 import { InputFields } from '../inputFields';
 import { Button } from '../button';
 import { Warning } from '../warning';
-import { login } from '../../services';
+import { login as loginRequest } from '../../services';
 import { NavLink } from 'react-router-dom';
```
Comment (arayik-99) [inline gta-gallery/src/components/login/index.jsx]: Merge this import with the upper import (2 line)

File: gta-gallery/src/components/editUser/index.jsx [hunk 22]
```diff
@@ -0,0 +1,58 @@
+import React, { useState } from 'react';
+import { editUser } from '../../services/userService';
+import { useHistory } from 'react-router-dom';
```
Comment (arayik-99) [inline gta-gallery/src/components/editUser/index.jsx:2]: Move up

File: gta-gallery/src/services/axiosInstance.js [hunk 23]
```diff
@@ -19,4 +19,12 @@ axiosInstance.interceptors.request.use((config) => {
   return config;
 });
 
+axiosInstance.interceptors.response.use((config) => {
+  const token = localStorage.getItem('token');
+  if (token) {
```
Comment (arayik-99) [inline gta-gallery/src/services/axiosInstance.js]: remove curly braces

**Review by arayik-99 — CHANGES_REQUESTED** [review]

File: gta-gallery/src/components/gtaVersionDetails/index.jsx [hunk 24]
```diff
@@ -25,6 +32,59 @@ export const GTAVersionDetails = () => {
...
+    if (id && imageValue && nameValue && informationValue) {
+      let newVersions = gtaVersionDetails;
+      newVersions = {
+        image: imageValue,
+        name: nameValue,
+        information: informationValue,
+      };
+      setGTAVersionDetails(newVersions);
+      await updateVersion(id, imageValue, nameValue, informationValue);
+      clearState();
```
Comment (Karen-Avdalyan) [inline gta-gallery/src/components/gtaVersionDetails/index.jsx]: you can call this method in closeModal function

File: gta-gallery/src/components/gtaVersionDetails/index.jsx [hunk 25]
```diff
@@ -25,6 +32,59 @@ export const GTAVersionDetails = () => {
...
+    },
+  ];
+
+  const closeUpdateModal = () => {
+    setUpdateModalVisible(!updateModalVisible);
+  };
+
+  const handleUpdate = async () => {
+    if (id && imageValue && nameValue && informationValue) {
+      let newVersions = gtaVersionDetails;
```
Comment (Karen-Avdalyan) [inline gta-gallery/src/components/gtaVersionDetails/index.jsx]: this row is redundant

File: gta-gallery/src/components/shared/addVersionModal/addVersionModal.jsx [hunk 26]
```diff
@@ -1,63 +1,19 @@
...
-      createVersion(image, nameValue, informationValue);
-      closeAddModal();
-    }
-    setWarning(true);
-  };
-
+export const AddVersionModal = ({
+  isOpen,
+  onCancel,
+  onOk,
```
Comment (Karen-Avdalyan) [inline gta-gallery/src/components/shared/addVersionModal/addVersionModal.jsx]: this prop name is not correct I think

File: gta-gallery/src/components/shared/updateVersionModal/updateVersionModal.jsx [hunk 27]
```diff
@@ -1,63 +1,19 @@
...
-      updateVersion(id, image, nameValue, informationValue);
-      closeUpdateModal();
-    }
-    setWarning(true);
-  };
-
+export const UpdateVersionModal = ({
+  isOpen,
+  onCancel,
+  onOk,
```
Comment (Karen-Avdalyan) [inline gta-gallery/src/components/shared/updateVersionModal/updateVersionModal.jsx]: this prop name is not correct I think

**Review by Karen-Avdalyan — CHANGES_REQUESTED** [review]

File: gta-gallery/src/layout/index.jsx [hunk 28]
```diff
@@ -1,7 +1,7 @@
 import React, { useState, useEffect } from 'react';
 import { getAllVersions, createVersion, deleteVersion } from '../services';
```
Comment (arayik-99) [inline gta-gallery/src/layout/index.jsx]: move this at the end of imports. Make sure everywhere it's imported the same.

File: gta-gallery/src/components/modal.jsx [hunk 29]
```diff
@@ -9,14 +8,14 @@ export const Modal = ({
   onCancel,
   onOk,
   header,
-  warning,
+  imageValue = '',
```
Comment (arayik-99) [inline gta-gallery/src/components/modal.jsx:11]: Why do we have default value here?

**Review by arayik-99 — CHANGES_REQUESTED** [review]

File: gta-gallery/src/components/modal.jsx (hunk 29 above)
Comment (SergeyGhazaryan) [reply gta-gallery/src/components/modal.jsx:11]: To show a picture when needed.

**Review by arayik-99 — APPROVED** [review]

**Pull request merged**